doc/dev/transforms.rst

```diff
@@ -0,0 +1,574 @@
+  argument of our operation with the result of the multiplication. Since this mechanism doesn't
+  go through the rewriter, he have to notify it explicitly that we are making changes to an
+  operation:
+
+  .. code-block:: cpp
+
+      rewriter.updateRootInPlace(op, [&] {
+          op->setOperand(0, res);
+      });
+
+  Note that in order to change to results on an operation you will need to create a copy of it
+  and erase the existing operation, they cannot be modified in-place.
+
+Invoking transformation patterns
+================================
+
+IR changes are always effected by a transformation *pass*. Many compilers are structured around the
+notion of passes, where the program is progressively transformed and each pass is responsible for a
+particular sub-task.
+
+While the transformation pattern we wrote above defines how we want to transform certain aspects of
+our program, it doesn't yet specify how the patterns are applied to an input program. For this we
+need to write a pass.
+
+The simplest approach might be to say we want our transformation pass to look at the entire program,
+and apply a set of patterns we defined like the one above. We can do so by creating an
+`OperationPass <https://mlir.llvm.org/docs/PassManagement/#operation-pass-static-filtering-by-op-type>`_
+that acts on an MLIR module (remember an MLIR module is an operation that itself contains globals
+and other function operations, which themselves can contain other operations, and so on):
+
+.. code-block:: cpp
+
+    struct QuantumOptimizationPass : public PassWrapper<QuantumOptimizationPass, OperationPass<ModuleOp>>
+    {
+        void runOnOperation() {
+            // Get the current operation being operated on.
+            ModuleOp op = getOperation();
+            MLIRContext *ctx = &getContext();
+
+            // Define the set of patterns to use.
+            RewritePatternSet quantumPatterns(ctx);
+            quantumPatterns.add<QubitUnitaryFusion>(ctx);
+
+            // Apply patters in an iterative and greedy manner.
+            if (failed(applyPatternsAndFoldGreedily(op, std::move(quantumPatterns)))) {
+                return signalPassFailure();
+            }
+        }
+    };
+
+To apply patterns we need a `pattern applicator <https://mlir.llvm.org/docs/PatternRewriter/#common-pattern-drivers>`_.
+There a few in MLIR but typically you can just use the greedy pattern rewrite driver
+(``applyPatternsAndFoldGreedily``), which will iterative over the IR and apply patterns until a
+fixed point is reached.
+
+
+Writing more general transformations
+====================================
+
+The pattern-based approach to transformations is not limited to small peephole optimizations like
+the one above, in fact all transformation passes in Catalyst currently use either regular rewrite
+patters or dialect conversion patterns. Let's take a quick look at the finite difference method
```

Comment:
Is it 'finite-difference method' (with the hyphen)? I can't work it out

Reply:
Good question, my guess is you are likely to find both out in the wild. Also is it "finite difference" or "finite differences"?

Reply:
I looked at wikipedia and they use both 😆 I think it might be finite-difference method, but method of finite differences.

Reply:
Thanks :D

Reply:
Updated all occurrences: https://github.com/PennyLaneAI/catalyst/pull/340/commits/2d66f3a67a4e0e0feef868aa44ff4fe7b14d5bbb